### PR TITLE
PR for full integration tests

### PR DESCRIPTION
**Pull request by xqjibz**
@guyellis can you take a look at this, and see if it's what you're thinking?

### CODE REVIEW

**Comment by xqjibz** [comment]
I was thinking [this](https://github.com/guyellis/multi-level-cache/blob/robust-testing/test/lib/cache-lib/test.redis.integration.js), then you'd run the test with `MULTICACHE_USE_REDIS=1 npm test` we can work out the names and the values to keep them more in line with what may already be out there.

File: lib/helpers/redis-stubs.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  'del': sinon.stub().callsArg(1),
+  'set': sinon.stub().callsArg(2)
+};
+
+var redisCreateClientStub = sinon.stub(redis, 'createClient', function(){
+  return clientStub;
+});
+
+exports.redisCreateClientStub = redisCreateClientStub;
+exports.clientStub = clientStub;
```
Comment (guyellis) [inline lib/helpers/redis-stubs.js]: If we were to use this file it shouldn't be in this location. Sinon is a dev dependency and this location is for production code. It should be in the test folder.

File: test/lib/cache-lib/test.redis.integration.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+//var redis = require('redis');
+var assert = require('assert');
+// conditional test.
+var runTest = false;
+
+if(Number(process.env.MULTICACHE_USE_REDIS)  === 1){
+  runTest = true;
+}
```
Comment (guyellis) [inline test/lib/cache-lib/test.redis.integration.js]: Lines 6 to 10 could be squashed into 1:
`+var runTest = Number(process.env.MULTICACHE_USE_REDIS) === 1;`

**Comment by guyellis** [comment]
I'm thinking along the lines of a single unit test file, as we had originally, that does the tests by mocking out the call to redis and always assumes that the user does not have it running.

The the integration test is run through the test.multi.js file. Currently this file has 'node-cache' "hard coded" as the local and remote cache. We can swap this out for redis in the integration test as this will exercise all the logic that we're expecting to be working.

The only thing that we need to test in the redis specific test is that the wrapping code works as we would expect. For example, if we were to allow an array of keys for the `del()` function and redis did not allow deleting by array (I don't know if it does) then we'd loop in the redis wrapper and call del() as many times as there are items in the array. This would not need to be integration tested via the redis tests as that functionality will be in the test.multi.js file and would be tested through there.

File: lib/helpers/redis-stubs.js (hunk 1 above)
Comment (xqjibz) [inline lib/helpers/redis-stubs.js]: @guyellis I tried to put this inside the test directory, but couldn't find a way to exclude it, as it's not a test.  Is there a way to exclude files from the basic `npm test` when that fires the `mocha` runner.  There are some options inside mocha itself, but they only deal with the test name.

Comment (guyellis) [inline lib/helpers/redis-stubs.js]: So long as you don't have a `describe()` inside the file I don't think that `mocha` will run it. i.e. it has nothing to execute so will implicitly ignore it.

**Pull request merged**